### ISSUE TITLE
Connect wp-cli to LEMP stack

### ISSUE THREAD

**Issue by robots4life** [docker-library/wordpress#366]
New to Docker. Got a working LEMP stack going. Trying to connect WP-CLI to that LEMP stack but no matter what I try WP-CLI does not want to connect or work.

```
version: "3.7"

services:
  php:
    build: 
      context: './config/php/'
    networks:
      - devnet
    volumes:
      - /media/user/d/WWW/:/var/www/html/
    working_dir: /var/www/html/
    container_name: php

  nginx:
    image: nginx:1.14-alpine
    depends_on:
      - php
      - mysql
    networks:
      - devnet      
    ports:
      - "80:80"
      - "443:443"
    volumes:
      - ./config/nginx/nginx.conf:/etc/nginx/conf.d/default.conf:ro
      - /media/user/d/WWW/:/var/www/html/
    container_name: nginx

  mysql:
    image: mysql:5.7
    restart: always
    ports:
      - "3306:3306"
    volumes:
      - /media/user/d/MYSQL/:/var/lib/mysql
    networks:
      - devnet      
    environment:
      MYSQL_ROOT_PASSWORD: "db_root"
      MYSQL_DATABASE: "test_db"
      MYSQL_USER: "test_db_user"
      MYSQL_PASSWORD: "test_db_pass"
    container_name: mysql

  wpcli:
    image: wordpress:cli
    depends_on:
      - mysql
    environment:
      # same as mysql/db service
      WORDPRESS_DB_HOST: mysql 
      # same as MYSQL_DATABASE
      WORDPRESS_DB_NAME: test_db
      # same as MYSQL_USER
      WORDPRESS_DB_USER: test_db_user
      # same as MYSQL_PASSWORD
      WORDPRESS_DB_PASSWORD: test_db_pass
    networks:
      - devnet
    volumes:
      - /media/user/d/WWW/:/var/www/html/
    working_dir: /var/www/html/
    container_name: wpcli

networks:
  devnet:
```

The PHP image is built from this Dockerfile.

```
FROM php:7.2-fpm-alpine

# docker-entrypoint.sh dependencies
RUN apk add --no-cache \
# in theory, docker-entrypoint.sh is POSIX-compliant, but priority is a working, consistent image
		bash \
# BusyBox sed is not sufficient for some of our sed expressions
		sed

# https://github.com/docker-library/wordpress/blob/master/php7.3/fpm-alpine/Dockerfile
# install the PHP extensions we need
RUN set -ex; \
	\
	apk add --no-cache --virtual .build-deps \
		libjpeg-turbo-dev \
		libpng-dev \
		libzip-dev \
	; \
	\
	docker-php-ext-configure gd --with-png-dir=/usr --with-jpeg-dir=/usr; \
	docker-php-ext-install gd mysqli opcache zip; \
	\
	runDeps="$( \
		scanelf --needed --nobanner --format '%n#p' --recursive /usr/local/lib/php/extensions \
			| tr ',' '\n' \
			| sort -u \
			| awk 'system("[ -e /usr/local/lib/" $1 " ]") == 0 { next } { print "so:" $1 }' \
	)"; \
	apk add --virtual .wordpress-phpexts-rundeps $runDeps; \
    apk del .build-deps
```

The NGINX conf is this.

```
upstream php {  
  server php:9000;
}

server {
        # http://nginx.org/en/docs/http/configuring_https_servers.html#single_http_https_server
        listen                  80;
        listen                  443 ssl;
        # listen                  [::]:443 ssl http2;
        
        server_name             localhost;

        ssl_certificate         /etc/nginx/certs/localhost-crt.pem;
        ssl_certificate_key     /etc/nginx/certs/localhost-key.pem;

        ssl_protocols           TLSv1.2 TLSv1.1 TLSv1;        


        ## Your only path reference.
        root /var/www/html;
        ## This should be in your http block and if it is, it's not needed here.
        index index.php index.html index.htm;

        location = /favicon.ico {
                log_not_found off;
                access_log off;
        }

        location = /robots.txt {
                allow all;
                log_not_found off;
                access_log off;
        }

        location / {
                # https://nginxlibrary.com/enable-directory-listing/
                autoindex on;

                # This is cool because no php is touched for static content.
                # include the "?$args" part so non-default permalinks doesn't break when using query string
                try_files $uri $uri/ /index.php?$args;
        }

        location ~ \.php$ {
                #NOTE: You should have "cgi.fix_pathinfo = 0;" in php.ini
                include fastcgi.conf;
                fastcgi_intercept_errors on;
                fastcgi_pass php;
        }

        location ~* \.(js|css|png|jpg|jpeg|gif|ico)$ {
                expires max;
                log_not_found off;
        }
}
```

Inside a WWW folder I have all work projects. Each folder either has WordPress or other projects. If I download the WordPress files into a subfolder of WWW I can run the install just fine. Everything works. It connects to the MYSQL service.

So if I go to e.g. `https://localhost/clientA/` the LEMP stack will serve the project from the folder clientA and use the MYSQL db service provided. If I just go to `https://localhost/` it will show me an index of all projects, each being inside a folder. I click on a folder and the project comes up.

What I would like to be able to do is access WP-CLI under that WWW folder, so make it work so that I can be in /var/www/html/ and type commands like `wp core download` or `wp config create  -dbname=wordpress_test_db --dbuser=wp_test_db_user` or `wp db create` and finally `wp core install ` to be able to just make a new WordPress site.

I also tried adding WP-CLI to the PHP service directly. Sure, downloading and install it works, however as soon as it needs to connect to the MYSQL service it can't. Before using Docker I was on a XAMPP portable dev environment. Ran WP-CLI, had it available system wide and could fire up a new WP site in seconds.

Do I really need to include a WordPress image to be able to run WP-CLI in a Docker environment with a LEMP stack?

I also set the environment variables of the WP-CLI service to the same variable the MYSQL service uses but it does not work.

Also when I pass this command
```
command: '--path=`/var/www/html/WP-Project/`'
```
to the WP-CLI service it exits, if I do not pass that command it tells me that under the current path no WordPress install could be found. But under WWW/WP-Project/ I have a working WordPress install and the db is connected to the MYSQL service.

Oh and yes, sure I thought of installing the entire development environment into one image with all the services inside that one image, but I read that this is against Docker's advised usage, wanting to keep services apart from each other.

Any help or guidance towards this would be great. 

```
Docker version 18.09.1, build 4c52b90
Distributor ID: Ubuntu
Description:    Ubuntu 18.04.1 LTS
Release:        18.04
Codename:       bionic
```

**Comment by wglambert** [comment]
https://github.com/docker-library/docs/tree/master/wordpress#wordpresscli
>This image variant does not contain WordPress itself, but instead contains [WP-CLI](https://wp-cli.org).

For better help with your setup you could ask [the Docker Community Forums](https://forums.docker.com/), [the Docker Community Slack](https://blog.docker.com/2016/11/introducing-docker-community-directory-docker-community-slack/), or [Stack Overflow](https://stackoverflow.com/search?tab=newest&q=docker). As these repositories are for issues with the image and not necessarily for questions of usability

**Comment by robots4life** [comment]
Anyone coming here or the places kindly linked above.

**Spare yourself the time, nerves and energy and use https://github.com/cytopia/devilbox**

It has all the tools, including wp-cli and works 100% out of the box, no frowning or loosing time on learning how to setup backend, services, etc. while you could be coding and getting on with it.

**Comment by demyxco** [comment]
@robots4life it's not even hard to install wp-cli. What're you doing wrong is including wordpress:cli in your docker-compose when it's supposed to be ran independently via docker run.

```
docker run -it --rm \
    --volumes-from wordpress_php_container \
    --network container:wordpress_php_container \
    wordpress:cli user list
```

This would create the wp-cli container, execute wp-cli commands to your wordpress_php_container, then self destruct.

**Comment by robots4life** [comment]
Too funny this comes 52 days later, given it is not even hard.
I recall getting it to work, albeit having to change the default mysql host to localhost in wp-config every time I fire up a new site. Local mysql server is on localhost and not on "mysql".
https://github.com/docker-library/wordpress/blob/master/php7.3/fpm/docker-entrypoint.sh#L139

During the meanwhilst I have dropped developing with WP completely.
After years of love the time has come to move on.
Nevertheless, thank you indeed for the input.
Fingers crossed this can be useful for others.

Though, if you struggle, give https://github.com/cytopia/devilbox a shot and continue coding full stop. Cannot stress enough how much time I have invested trying to get things running, just thinking about trying to learn what PHP modules are needed/included and then looking at the error messages when firing up the containers.
